README.md
# Lock Posts


**INACTIVE NOTICE: This plugin is unsupported by WPMUDEV, we've published it here for those technical types who might want to fork and maintain it for their needs.**

## Translations

Translation files can be found at https://github.com/wpmudev/translations

## About

Lock Posts lets a network super admin 'lock' or 'fix' posts so they can't be edited by the users of that site – even administrators.

### Block post editing

Perfect for stopping school assignments from being edited after submission or adding a post to a network that you don't want removed. Lock Posts adds a post status module that's only available to post or page admins. Easily toggle the status between locked and unlocked. 

![Lock Posts Editor](http://premium.wpmudev.org/wp-content/uploads/2009/06/lock-posts-1150-editor.png)

 Quickly toggle between lock status

 Quickly see what posts have been locked by the super admin from 'All Posts'. 

![Lock Posts Locked](http://premium.wpmudev.org/wp-content/uploads/2009/06/lock-posts-1150-locked.png)

 Status information conveniently shown in 'All Posts'

 Take control of what posts can be edited across an entire Multisite network with Lock Posts.

## Usage

### To Get Started:

Start by reading the [Installing Plugins](https://premium.wpmudev.org/wpmu-manual/installing-regular-plugins-on-wpmu/) section in our comprehensive [WordPress and WordPress Multisite Manual](https://premium.wpmudev.org/wpmu-manual/) if you are new to WordPress. _Important notes:_

*   If you have an older version of the plugin installed in /mu-plugins/ please delete it.
*   When this plugin is activated on a single site, only admins have the capability to lock/unlock posts.
*   When activated on any site in a multisite install, it is activated on all sites, but only the network admin can lock/unlock posts.

### To Use:

Log into your site as an admin (or network-admin) user. Then go to any post, page or custom post type you want to lock. Click the "Edit" link and scroll to the bottom of the edit screen. 

![Lock Posts Editor](https://premium.wpmudev.org/wp-content/uploads/2009/06/lock-posts-1150-editor.png)

 Once a post has been locked, you'll see that indicated in a new column on the All Posts screen. 

![Lock Posts Locked](https://premium.wpmudev.org/wp-content/uploads/2009/06/lock-posts-1150-locked.png)

 Any other user who tries to edit a locked post will get a friendly message informing them that it has been locked. 

![Lock Posts Locked Message](https://premium.wpmudev.org/wp-content/uploads/2009/06/lock-posts-1150-locked-message.png)

 Please note the following:

*   On a network install, only super-admin can lock/unlock posts. Individual site admins and other roles cannot.
*   On a single site install, site admin can lock/unlock, other roles cannot.


changelog.txt
Plugin Name: Lock Posts
Author: WPMUDEV
Lead developer: Maniu
Contributors: Andrew Billits, Ulrich Sossou

Change Log:
----------------------------------------------------------------------
1.1.7 - 12/16/2016
----------------------------------------------------------------------

- Updated WPMU DEV Dashboard Notification

1.1.6 - 11/13/2015
----------------------------------------------------------------------
- Update dash notice


1.1.5 - 11/13/2013
----------------------------------------------------------------------
- Added support for all public custom post types.

1.1 - 08/10/2012
----------------------------------------------------------------------
- Locking based on user capability querying interception.
- Filterable, post type based scope locking.
- Filterable set of locking capabilities.

1.0.3 - 01/10/2012
----------------------------------------------------------------------
- Fix for submenu checks.

1.0.2 - 01/04/2011
----------------------------------------------------------------------
- 3.1+ compatibility.
- Internationalization.

1.0.1 - 08/31/2009
----------------------------------------------------------------------
- Fixed bug that resulted in posts getting locked regardless of setting.

1.0.0 - 06/19/2009
----------------------------------------------------------------------
- Initial Release.

locked.php
<?php
/*
Plugin Name: Lock Posts
Plugin URI: 
Description:
Author: Andrew Billits (Incsub)
Version: 1.0.1
Author URI: http://incsub.com
*/

/* 
Copyright 2007-2009 Incsub (http://incsub.com)

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License (Version 2 - GPLv2) as published by
the Free Software Foundation.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

require_once('admin.php');
$title = __('Post Locked');
$parent_file = 'locked.php';
require_once('admin-header.php');

lock_posts_locked();

include('admin-footer.php');
?>

lock-posts.php
<?php
/*
Plugin Name: Lock Posts
Plugin URI: https://premium.wpmudev.org/project/lock-posts/
Description: This plugin allows site admin to lock down posts on any blog so that regular ol' users just can't edit them - for example, with a school assignment - stop it from being edited after submission.
Author: WPMU DEV
Version: 1.1.7
Text Domain: lock_posts
Author URI: http://premium.wpmudev.org/
WDP ID: 83
*/

/*
Copyright 2007-2015 Incsub (http://incsub.com)

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License (Version 2 - GPLv2) as published by
the Free Software Foundation.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

class Lock_Posts {

	var $post_types;
	var $lock_capabilities;

	/**
	 * PHP5 constructor
	 *
	 */
	function __construct() {

		$this->lock_capabilities = array(
			'edit_post',
			'delete_post',
		);

		add_action( 'add_meta_boxes', array( &$this, 'meta_box' ) );
		add_action( 'admin_menu', array( &$this, 'admin_page' ) );
		add_action( 'save_post', array( &$this, 'update' ) );
		add_action( 'init', array( &$this, 'check' ) );

		add_filter('user_has_cap', array($this, 'kill_edit_cap'), 10, 3);

		add_action( 'init', array( &$this, 'add_columns' ), 99 );

		add_action( 'plugins_loaded', array( &$this, 'load_languages' ) );
	}

	function load_languages() {
		// load text domain
		if ( defined( 'WPMU_PLUGIN_DIR' ) && file_exists( WPMU_PLUGIN_DIR . '/lock-posts.php' ) ) {
			load_muplugin_textdomain( 'lock_posts', 'lock-posts-files/languages' );
		} else {
			load_plugin_textdomain( 'lock_posts', false, dirname( plugin_basename( __FILE__ ) ) . '/lock-posts-files/languages' );
		}
	}

	/**
	 * Properly filtering out forbidden capabilities for non-super admin users.
	 */
	function kill_edit_cap ($all, $caps, $args) {
		global $post;
		if (!is_object($post) || !isset($post->post_type)) return $all; // Only proceed for pages with known post types
		if (!in_array($post->post_type, $this->lock_capabilities)) return $all; // Only proceed for allowed types

		if (!$args) return $all; // Something is wrong here.
		if (count($args) < 3) return $all; // Only proceed for individual items.
		if (!isset($args[0])) return $all; // Something is still wrong here.

		//if ('edit_post' != $args[0]) return $all; // So, not the one we're looking for.
		if (!in_array($args[0], $this->lock_capabilities)) return $all; // Not one of the forbidden capabilities
		$post_id = isset($args[2]) ? $args[2] : false;
		if (!$post_id) return $all; // Can't obtain post ID

		$post_lock_status = ('locked' == get_post_meta($post_id, '_post_lock_status', true));

		return $post_lock_status ? (is_super_admin() ? $all : false) : $all;
	}

	/**
	 * Properly manage columns
	 *
	 */
	function add_columns() {
		$this->post_types = get_post_types(array('show_ui' => true, 'public' => true));
		unset($this->post_types['attachment']);

		foreach ($this->post_types as $post_type) {
			add_filter( 'manage_edit-'.$post_type.'_columns', array( &$this, 'status_column' ), 999 );
			add_action( 'manage_'.$post_type.'_posts_custom_column', array( &$this, 'status_output' ), 99, 2 );
		}
	}

	/**
	 * Add column to posts management panel
	 *
	 */
	function status_column( $columns ) {
		$columns['lock_status'] = __( 'Lock Status', 'lock_posts' );
		return $columns;
	}

	/**
	 * Display columns content on posts management panel
	 *
	 */
	function status_output( $column, $id ) {
		if( $column == 'lock_status' ) {
			$post_lock_status = get_post_meta( $id, '_post_lock_status' );

			if( is_array( $post_lock_status ) && isset( $post_lock_status[0] ) )
				$post_lock_status = $post_lock_status[0];

			if( 'locked' == $post_lock_status )
				echo __( 'Locked', 'lock_posts' );
			else
				echo __( 'Unlocked', 'lock_posts' );
		}
	}

	/**
	 * Add metabox to posts edition panel
	 *
	 */
	function meta_box() {
		if (!is_super_admin()) return;

		foreach ($this->post_types as $key => $post_type) {
			add_meta_box( 'postlock', __( 'Post Status', 'lock_posts' ), array( &$this, 'meta_box_output' ), $post_type, 'advanced', 'high' );
		}
	}

	/**
	 * Post status metabox
	 *
	 */
	function meta_box_output( $post ) {
		if ( !is_super_admin() )
			return;

		$post_lock_status = get_post_meta( $post->ID, '_post_lock_status' );
		if( is_array( $post_lock_status ) && isset( $post_lock_status[0] ) )
			$post_lock_status = $post_lock_status[0];

		if( empty( $post_lock_status ) )
			$post_lock_status = 'unlocked';
		?>
		<div id="postlockstatus">
			<label class="hidden" for="excerpt">Post Status</label>
			<select name="post_lock_status">
				<option value="locked" <?php selected( $post_lock_status, 'locked' ) ?>><?php _e( 'Locked', 'lock_posts' ) ?></option>
				<option value="unlocked" <?php selected( $post_lock_status, 'unlocked' ) ?>><?php _e( 'Unlocked', 'lock_posts' ) ?></option>
			</select>
			<p><?php _e( 'Locked posts cannot be edited by anyone other than Super admins.', 'lock_posts' ); ?></p>
		</div>
		<?php
	}

	/**
	 * Update post status
	 *
	 */
	function update( $post_id ) {
		if ( !empty( $_POST['post_lock_status'] ) && is_super_admin() )
			update_post_meta( $post_id, '_post_lock_status', $_POST['post_lock_status'] );
	}

	/**
	 * Check post status and redirect if the user is not super admin and post is locked
	 *
	 */
	function check() {
		if ( !is_super_admin() && !empty( $_GET['action'] ) && 'edit' == $_GET['action'] && !empty( $_GET['post'] ) ) {
			$post_lock_status = get_post_meta( $_GET['post'], '_post_lock_status' );

			if ( is_array($post_lock_status) )
				$post_lock_status = $post_lock_status[0];

			if ( $post_lock_status == 'locked' )
				wp_redirect( admin_url( 'edit.php?page=post-locked&post=' . $_GET['post'] ) );
		}
	}

	/**
	 * Displayed 'locked' message
	 *
	 */
	function locked() {
		$post = get_post( $_GET['post'] );
		echo '<div class="wrap">';
		echo '<h2>' . __( 'Post Locked', 'lock_posts' ) . '</h2>';
		echo '<p>' . sprintf( __( 'The post "%s" has been locked by a Super admin and you aren\'t able to edit it.', 'lock_posts' ), $post->post_title ) . '</p>';
		echo '<p><a href="' . admin_url( 'edit.php?post_type=' . $post->post_type ) . '">&laquo; ' . __( 'Back to Posts List', 'lock_posts' ) . '</a></p>';
		echo '</div>';
	}

	/**
	 * Add admin page
	 *
	 */
	function admin_page() {
		global $submenu;

		add_submenu_page( 'edit.php', 'Post Locked', 'Post Locked', 'edit_posts', 'post-locked', array( &$this, 'locked' ) );

		if (isset($submenu['edit.php']) && is_array($submenu['edit.php'])) foreach( $submenu['edit.php'] as $key => $menu_item ) {
			if( isset( $menu_item[2] ) && $menu_item[2] == 'post-locked' )
				unset( $submenu['edit.php'][$key] );
		}
	}

}

$lock_posts = new Lock_Posts();

global $wpmudev_notices;
$wpmudev_notices[] = array( 'id'=> 83, 'name'=> 'Lock Posts', 'screens' => array( 'post' ) );
include_once(plugin_dir_path( __FILE__ ).'dash-notice/wpmudev-dash-notification.php');

